routing/index_graph.cpp

```diff
@@ -13,27 +16,520 @@ IndexGraph::IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimat
+  RoadGeometry const & road = GetRoad(rp.GetFeatureId());
+  CHECK_LESS(rp.GetPointId(), road.GetPointsCount(), ());
+  return road.GetPoint(rp.GetPointId());
+}
+
 m2::PointD const & IndexGraph::GetPoint(Joint::Id jointId)
 {
-  return m_geometry.GetPoint(m_jointIndex.GetPoint(jointId));
+  return GetPoint(m_jointIndex.GetPoint(jointId));
 }
 
+double IndexGraph::GetSpeed(RoadPoint const & rp) { return GetRoad(rp.GetFeatureId()).GetSpeed(); }
+
 void IndexGraph::Build(uint32_t numJoints) { m_jointIndex.Build(m_roadIndex, numJoints); }
 
 void IndexGraph::Import(vector<Joint> const & joints)
 {
-  m_roadIndex.Import(joints);
   CHECK_LESS_OR_EQUAL(joints.size(), numeric_limits<uint32_t>::max(), ());
+  m_roadIndex.Import(joints);
   Build(static_cast<uint32_t>(joints.size()));
 }
 
+void IndexGraph::GetSingleFeaturePath(RoadPoint const & from, RoadPoint const & to,
+                                      vector<RoadPoint> & path)
+{
+  CHECK_EQUAL(from.GetFeatureId(), to.GetFeatureId(), ());
+
+  path.clear();
+  int const shift = to.GetPointId() > from.GetPointId() ? 1 : -1;
+  for (uint32_t i = from.GetPointId(); i != to.GetPointId(); i += shift)
+    path.emplace_back(from.GetFeatureId(), i);
+  path.push_back(to);
+}
+
+void IndexGraph::GetConnectionPaths(Joint::Id from, Joint::Id to,
+                                    vector<vector<RoadPoint>> & connectionPaths)
+{
+  CHECK_NOT_EQUAL(from, Joint::kInvalidId, ());
+  CHECK_NOT_EQUAL(to, Joint::kInvalidId, ());
+
+  connectionPaths.clear();
+  vector<pair<RoadPoint, RoadPoint>> connections;
+  m_jointIndex.FindPointsWithCommonFeature(from, to, connections);
+  if (connections.empty())
+    return;
+
+  connectionPaths.reserve(connections.size());
+  for (auto const & c : connections)
+  {
+    connectionPaths.emplace_back();
+    GetSingleFeaturePath(c.first /* from */, c.second /* to */, connectionPaths.back());
+  }
+}
+
+void IndexGraph::GetFeatureConnectionPath(Joint::Id from, Joint::Id to, uint32_t featureId,
+                                          vector<RoadPoint> & path)
+{
+  path.clear();
+  vector<pair<RoadPoint, RoadPoint>> connections;
+  m_jointIndex.FindPointsWithCommonFeature(from, to, connections);
+  if (connections.empty())
+    return;
+
+  for (auto & c : connections)
+  {
+    CHECK_EQUAL(c.first.GetFeatureId(), c.second.GetFeatureId(), ());
+    if (c.first.GetFeatureId() == featureId)
+    {
+      GetSingleFeaturePath(c.first /* from */, c.second /* to */, path);
+      return;
+    }
+  }
+}
+
+void IndexGraph::GetOutgoingGeomEdges(vector<JointEdge> const & outgoingEdges, Joint::Id center,
+                                      vector<JointEdgeGeom> & outgoingGeomEdges)
+{
+  for (JointEdge const & e : outgoingEdges)
+  {
+    vector<vector<RoadPoint>> connectionPaths;
+    GetConnectionPaths(center, e.GetTarget(), connectionPaths);
+    if (connectionPaths.empty())
+      MYTHROW(RoutingException, ("Can't find common feature for joints", center, e.GetTarget()));
```

Comment:
I put LOG(LERROR, ("Can't find common feature for joints", center, e.GetTarget())); in that case.